geopolars/src/spatial_index.rs

```diff
@@ -0,0 +1,621 @@
+            (Geometry::MultiPolygon(poly), Geometry::Point(point), _) => poly.contains(point),
+
+            // Polygon and Polygon
+            (Geometry::Polygon(poly_lhs), Geometry::Polygon(poly_rhs), Predicate::Contains) => {
+                poly_lhs.contains(poly_rhs)
+            }
+            (Geometry::Polygon(poly_lhs), Geometry::Polygon(poly_rhs), Predicate::Intersects) => {
+                poly_lhs.intersects(poly_rhs)
+            }
+
+            // Multi Polygon and Polygon
+            (
+                Geometry::MultiPolygon(poly_lhs),
+                Geometry::Polygon(poly_rhs),
+                Predicate::Contains,
+            ) => poly_lhs.contains(poly_rhs),
+            (
+                Geometry::MultiPolygon(poly_lhs),
+                Geometry::Polygon(poly_rhs),
+                Predicate::Intersects,
+            ) => poly_lhs.intersects(poly_rhs),
+
+            // Polygon and MultiPolygon
+            (
+                Geometry::Polygon(poly_lhs),
+                Geometry::MultiPolygon(poly_rhs),
+                Predicate::Intersects,
+            ) => poly_lhs.intersects(poly_rhs),
+
+            // Line and Point
+            (Geometry::Line(line), Geometry::Point(point), _) => line.contains(point),
+            (Geometry::Point(point), Geometry::Line(line), _) => line.contains(point),
+
+            // LineString and Point
+            (Geometry::LineString(line), Geometry::Point(point), _) => line.contains(point),
+            (Geometry::Point(point), Geometry::LineString(line), _) => line.contains(point),
+
+            // MultiLineString and Point
+            (Geometry::MultiLineString(line), Geometry::Point(point), _) => line.contains(point),
+            (Geometry::Point(point), Geometry::MultiLineString(line), _) => line.contains(point),
+            _ => false,
+        };
+
+        if actual_hit {
+            left_series.push(lhs_node.index);
+            right_series.push(rhs_node.index);
+        }
+    }
+
+    // Now we have two vecs with the alligned left right node indexes we perform a
+    // join using polars existing code.
+    let lhs_index: Vec<u64> = (0..lhs.shape().0).map(|i| i as u64).collect();
+    let rhs_index: Vec<u64> = (0..rhs.shape().0).map(|i| i as u64).collect();
+
+    let lhs_index = Series::new("lhs_index", lhs_index);
+    let rhs_index = Series::new("rhs_index", rhs_index);
+
+    let lhs_join_series: Vec<u64> = left_series.iter().map(|i| *i as u64).collect();
+    let rhs_join_series: Vec<u64> = right_series.iter().map(|i| *i as u64).collect();
+
+    let lhs_join_series = Series::new("lhs_join", lhs_join_series);
+    let rhs_join_series = Series::new("rhs_join", rhs_join_series);
+
+    let join_df: DataFrame = DataFrame::new(vec![lhs_join_series, rhs_join_series])?;
+
+    let mut lhs_with_index = lhs.hstack(&[lhs_index])?;
+    let mut rhs_with_index = rhs.hstack(&[rhs_index])?;
```

Comment:
What happens here if the left df already has a column named `lhs_join` or the right df already has a column `rhs_join`?